File: server/app/routes/discover.py
from fastapi import APIRouter, Depends, HTTPException, status, Query
from sqlalchemy.orm import Session
from sqlalchemy import text, func, or_, and_
from typing import List, Optional
from ..database import get_db
from ..models import Flashcard, Story, Image, Translation
from ..schemas import Flashcard as FlashcardSchema, Story as StorySchema, Image as ImageSchema, Translation as TranslationSchema
from datetime import datetime, timedelta

router = APIRouter(prefix="/discover", tags=["discover"])

# Age group mapping
AGE_GROUPS = {
    "toddler": (1, 3),
    "preschool": (4, 5),
    "elementary": (6, 10),
    "middle_school": (11, 13)
}

@router.get("/")
async def discover_content(
    search: Optional[str] = Query(None, description="Search term for original word, translation, or story content"),
    language: Optional[str] = Query(None, description="Filter by target language"),
    age_group: Optional[str] = Query(None, description="Filter by age group (toddler, preschool, elementary, middle_school)"),
    content_type: Optional[str] = Query(None, description="Filter by content type (flashcards, stories, images, translations)"),
    skip: int = Query(0, ge=0, description="Number of items to skip"),
    limit: int = Query(20, ge=1, le=100, description="Number of items to return"),
    db: Session = Depends(get_db)
):
    """
    Discover public content with search and filtering capabilities.
    Returns a unified feed of all content types in descending order by creation time.
    """
    try:
        content_items = []
        
        # Get flashcards
        if not content_type or content_type == "flashcards":
            flashcard_query = db.query(Flashcard)
            
            # Apply search filter
            if search:
                search_term = f"%{search}%"
                flashcard_query = flashcard_query.filter(
                    or_(
                        Flashcard.original_word.ilike(search_term),
                        Flashcard.translated_word.ilike(search_term)
                    )
                )
            
            # Apply language filter
            if language:
                flashcard_query = flashcard_query.filter(Flashcard.target_language == language)
            
            flashcards = flashcard_query.order_by(Flashcard.created_at.desc()).offset(skip).limit(limit).all()
            
            for flashcard in flashcards:
                content_items.append({
                    "id": flashcard.id,
                    "content_type": "flashcard",
                    "original_word": flashcard.original_word,
                    "translated_word": flashcard.translated_word,
                    "target_language": flashcard.target_language,
                    "example_sentences": flashcard.example_sentences,
                    "created_at": flashcard.created_at.isoformat() if flashcard.created_at else None,
                    "story_title": None,
                    "story_content": None,
                    "story_theme": None,
                    "target_age_range": None,
                    "image_url": None,
                    "translation": None,
                    "explanation": None,
                    "user": {
                        "email": flashcard.user.email if flashcard.user else None
                    }
                })
        
        # Get stories
        if not content_type or content_type == "stories":
            story_query = db.query(Story)
            
            # Apply search filter
            if search:
                search_term = f"%{search}%"
                story_query = story_query.filter(
                    or_(
                        Story.story_title.ilike(search_term),
                        Story.story_content.ilike(search_term)
                    )
                )
            
            # Apply language filter
            if language:
                story_query = story_query.filter(Story.target_language == language)
            
            # Apply age group filter
            if age_group and age_group in AGE_GROUPS:
                story_query = story_query.filter(Story.target_age_range == age_group)
            
            stories = story_query.order_by(Story.created_at.desc()).offset(skip).limit(limit).all()
            
            for story in stories:
                content_items.append({
                    "id": story.id,
                    "content_type": "story",
                    "original_word": None,
                    "translated_word": None,
                    "target_language": story.target_language,
                    "example_sentences": None,
                    "created_at": story.created_at.isoformat() if story.created_at else None,
                    "story_title": story.story_title,
                    "story_content": story.story_content,
                    "story_theme": story.story_theme,
                    "target_age_range": story.target_age_range,
                    "image_url": None,
                    "translation": None,
                    "explanation": None,
                    "user": {
                        "email": story.user.email if story.user else None
                    }
                })
        
        # Get images
        if not content_type or content_type == "images":
            image_query = db.query(Image).filter(Image.status == "completed")
            
            # Apply search filter
            if search:
                search_term = f"%{search}%"
                image_query = image_query.filter(
                    or_(
                        Image.original_word.ilike(search_term),
                        Image.translated_word.ilike(search_term)
                    )
                )
            
            # Apply language filter
            if language:
                image_query = image_query.filter(Image.target_language == language)
            
            images = image_query.order_by(Image.created_at.desc()).offset(skip).limit(limit).all()
            
            for image in images:
                content_items.append({
                    "id": image.id,
                    "content_type": "image",
                    "original_word": image.original_word,
                    "translated_word": image.translated_word,
                    "target_language": image.target_language,
                    "example_sentences": None,
                    "created_at": image.created_at.isoformat() if image.created_at else None,
                    "story_title": None,
                    "story_content": None,
                    "story_theme": None,
                    "target_age_range": None,
                    "image_url": image.image_url,
                    "translation": None,
                    "explanation": None,
                    "user": {
                        "email": image.user.email if image.user else None
                    }
                })
        
        # Sort all content by creation date (newest first)
        content_items.sort(key=lambda x: x["created_at"], reverse=True)
        
        # Apply pagination to the combined results
        total_items = len(content_items)
        paginated_items = content_items[skip:skip + limit]
        
        return {
            "items": paginated_items,
            "total": total_items,
            "skip": skip,
            "limit": limit,
            "has_more": skip + limit < total_items
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error retrieving content: {str(e)}"
        )

@router.get("/test")
async def test_discover(db: Session = Depends(get_db)):
    """Test endpoint to check database connectivity"""
    try:
        flashcard_count = db.query(func.count(Flashcard.id)).scalar()
        return {
            "status": "success",
            "flashcard_count": flashcard_count,
            "message": "Database connection working"
        }
    except Exception as e:
        return {
            "status": "error",
            "error": str(e),
            "message": "Database connection failed"
        }

@router.get("/trending")
async def get_trending_content(
    limit: int = Query(10, ge=1, le=50, description="Number of trending items to return"),
    db: Session = Depends(get_db)
):
    """
    Get trending content based on view count and recent activity.
    """
    try:
        # Get trending stories (by view count)
        trending_stories = db.query(Story).order_by(Story.view_count.desc(), Story.created_at.desc()).limit(limit).all()
        
        # Get recent content (last 7 days)
        recent_date = datetime.now() - timedelta(days=7)
        recent_flashcards = db.query(Flashcard).filter(Flashcard.created_at >= recent_date).order_by(Flashcard.created_at.desc()).limit(limit).all()
        recent_images = db.query(Image).filter(Image.created_at >= recent_date, Image.status == "completed").order_by(Image.created_at.desc()).limit(limit).all()
        
        # Combine and sort by creation date
        trending_items = []
        
        for story in trending_stories:
            trending_items.append({
                "id": story.id,
                "content_type": "story",
                "title": story.story_title,
                "target_language": story.target_language,
                "target_age_range": story.target_age_range,
                "view_count": story.view_count,
                "created_at": story.created_at.isoformat()
            })
        
        for flashcard in recent_flashcards:
            trending_items.append({
                "id": flashcard.id,
                "content_type": "flashcard",
                "title": flashcard.original_word,
                "target_language": flashcard.target_language,
                "created_at": flashcard.created_at.isoformat()
            })
        
        for image in recent_images:
            trending_items.append({
                "id": image.id,
                "content_type": "image",
                "title": image.original_word,
                "target_language": image.target_language,
                "image_url": image.image_url,
                "created_at": image.created_at.isoformat()
            })
        
        # Sort by creation date (newest first)
        trending_items.sort(key=lambda x: x["created_at"], reverse=True)
        
        return {
            "trending_items": trending_items[:limit]
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error retrieving trending content: {str(e)}"
        )

@router.get("/stats")
async def get_discovery_stats(db: Session = Depends(get_db)):
    """
    Get statistics about available content for discovery.
    """
    try:
        # Count content by type
        flashcard_count = db.query(func.count(Flashcard.id)).scalar()
        story_count = db.query(func.count(Story.id)).scalar()
        image_count = db.query(func.count(Image.id)).filter(Image.status == "completed").scalar()
        translation_count = db.query(func.count(Translation.id)).scalar()
        
        # Count by language
        language_stats = db.query(
            Flashcard.target_language,
            func.count(Flashcard.id).label("count")
        ).group_by(Flashcard.target_language).all()
        
        # Count by age group
        age_stats = db.query(
            Story.target_age_range,
            func.count(Story.id).label("count")
        ).filter(Story.target_age_range.isnot(None)).group_by(Story.target_age_range).all()
        
        return {
            "content_counts": {
                "flashcards": flashcard_count,
                "stories": story_count,
                "images": image_count,
                "translations": translation_count,
                "total": flashcard_count + story_count + image_count
            },
            "language_distribution": [
                {"language": lang, "count": count} for lang, count in language_stats
            ],
            "age_group_distribution": [
                {"age_group": age, "count": count} for age, count in age_stats
            ]
        }
        
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Error retrieving discovery stats: {str(e)}"
        ) 

File: client/src/stores/discover.ts
import { defineStore } from 'pinia';
import { ref, computed } from 'vue';

export interface DiscoverItem {
  id: number;
  content_type: 'flashcard' | 'story' | 'image' | 'translation';
  original_word?: string;
  translated_word?: string;
  target_language?: string;
  example_sentences?: string[];
  created_at?: string;
  story_title?: string;
  story_content?: string;
  story_theme?: string;
  target_age_range?: string;
  image_url?: string;
  translation?: string;
  explanation?: string;
  user?: {
    email?: string;
  };
}

export interface DiscoverFilters {
  search: string;
  language: string;
  age_group: string;
  content_type: string;
}

export interface DiscoverStats {
  content_counts: {
    flashcards: number;
    stories: number;
    images: number;
    translations: number;
    total: number;
  };
  language_distribution: Array<{ language: string; count: number }>;
  age_group_distribution: Array<{ age_group: string; count: number }>;
}

export const useDiscoverStore = defineStore('discover', () => {
  const items = ref<DiscoverItem[]>([]);
  const loading = ref(false);
  const error = ref<string | null>(null);
  const hasMore = ref(false);
  const total = ref(0);
  const skip = ref(0);
  const limit = ref(20);
  
  const filters = ref<DiscoverFilters>({
    search: '',
    language: '',
    age_group: '',
    content_type: ''
  });
  
  const stats = ref<DiscoverStats | null>(null);
  const trendingItems = ref<DiscoverItem[]>([]);

  const API_BASE = import.meta.env.VITE_API_BASE_URL || 'http://127.0.0.1:8000';

  // Reset pagination when filters change
  const resetPagination = () => {
    skip.value = 0;
    items.value = [];
  };

  // Fetch discovery content
  const fetchContent = async (reset = false) => {
    try {
      loading.value = true;
      error.value = null;
      
      if (reset) {
        resetPagination();
      }
      
      const params = new URLSearchParams({
        skip: skip.value.toString(),
        limit: limit.value.toString()
      });
      
      // Add filters
      if (filters.value.search) {
        params.append('search', filters.value.search);
      }
      if (filters.value.language) {
        params.append('language', filters.value.language);
      }
      if (filters.value.age_group) {
        params.append('age_group', filters.value.age_group);
      }
      if (filters.value.content_type) {
        params.append('content_type', filters.value.content_type);
      }
      
      const response = await fetch(`${API_BASE}/api/discover/?${params.toString()}`);
      
      if (!response.ok) {
        throw new Error('Failed to fetch discovery content');
      }
      
      const data = await response.json();
      
      if (reset) {
        items.value = data.items;
      } else {
        items.value.push(...data.items);
      }
      
      hasMore.value = data.has_more;
      total.value = data.total;
      skip.value += data.items.length;
      
    } catch (err: any) {
      error.value = err.message;
      throw err;
    } finally {
      loading.value = false;
    }
  };

  // Load more content (for infinite scroll)
  const loadMore = async () => {
    if (!hasMore.value || loading.value) return;
    await fetchContent(false);
  };

  // Update filters and refetch content
  const updateFilters = async (newFilters: Partial<DiscoverFilters>) => {
    Object.assign(filters.value, newFilters);
    await fetchContent(true);
  };

  // Clear all filters
  const clearFilters = async () => {
    filters.value = {
      search: '',
      language: '',
      age_group: '',
      content_type: ''
    };
    await fetchContent(true);
  };

  // Fetch trending content
  const fetchTrending = async () => {
    try {
      const response = await fetch(`${API_BASE}/api/discover/trending?limit=10`);
      
      if (!response.ok) {
        throw new Error('Failed to fetch trending content');
      }
      
      const data = await response.json();
      trendingItems.value = data.trending_items;
      
    } catch (err: any) {
      console.error('Failed to fetch trending content:', err);
    }
  };

  // Fetch discovery statistics
  const fetchStats = async () => {
    try {
      const response = await fetch(`${API_BASE}/api/discover/stats`);
      
      if (!response.ok) {
        throw new Error('Failed to fetch discovery stats');
      }
      
      stats.value = await response.json();
      
    } catch (err: any) {
      console.error('Failed to fetch discovery stats:', err);
    }
  };

  // Computed properties for filtered content
  const filteredItems = computed(() => {
    return items.value;
  });

  const contentByType = computed(() => {
    const grouped = {
      flashcards: items.value.filter(item => item.content_type === 'flashcard'),
      stories: items.value.filter(item => item.content_type === 'story'),
      images: items.value.filter(item => item.content_type === 'image'),
      translations: items.value.filter(item => item.content_type === 'translation')
    };
    return grouped;
  });

  const availableLanguages = computed(() => {
    if (!stats.value) return [];
    return stats.value.language_distribution.map(item => item.language);
  });

  const availableAgeGroups = computed(() => {
    if (!stats.value) return [];
    return stats.value.age_group_distribution.map(item => item.age_group);
  });

  // Initialize discovery
  const initialize = async () => {
    await Promise.all([
      fetchContent(true),
      fetchTrending(),
      fetchStats()
    ]);
  };

  return {
    // State
    items,
    loading,
    error,
    hasMore,
    total,
    filters,
    stats,
    trendingItems,
    
    // Actions
    fetchContent,
    loadMore,
    updateFilters,
    clearFilters,
    fetchTrending,
    fetchStats,
    initialize,
    
    // Computed
    filteredItems,
    contentByType,
    availableLanguages,
    availableAgeGroups
  };
}); 